File: Backend/models/userModel.js
import mongoose from 'mongoose'

const userSchema = new mongoose.Schema({
    name: { type: String, required: true },
    email: { type: String, required: true, unique: true },
    password: { type: String }, // Optional for mobile-only users, but required for email users
    phone: { type: String, unique: true, sparse: true }, // sparse allows multiple nulls if unique is true
    otp: { type: String, default: '' },
    otpExpireAt: { type: Number, default: 0 },
    verifyOtp: { type: String, default: '' },
    verifyOtpExpireAt: { type: Number, default: 0 },
    isAccountVerified: { type: Boolean, default: false },
    resetOtp: { type: String, default: '' },
    resetOtpExpireAt: { type: Number, default: 0 },
    role: { type: String, enum: ['user', 'admin', 'field-officer'], default: 'user' },
    district: { type: String, default: 'Other' },
    crops: { type: [String], default: [] },
    fieldOfficer: { type: mongoose.Schema.Types.ObjectId, ref: 'user', default: null },
    language: { type: String, default: 'en' },
    preferredLanguage: { type: String, default: 'en' },
    hasCompletedTour: { type: Boolean, default: false },
    simpleMode: { type: Boolean, default: false },
    cartData: { type: Object, default: {} },
    hasCompletedSurvey: { type: Boolean, default: false },
    surveyData: {
        language: { type: String },
        farmName: { type: String },
        farmSize: { type: Number },
        farmSizeUnit: { type: String, default: 'acres' },
        landOwnership: { type: String },
        soilType: { type: String },
        waterSource: { type: String },
        primaryCrops: [{ type: String }],
        farmingExperience: { type: String },
    },
    addresses: [{
        fullName: { type: String, required: true },
        phone: { type: String, required: true },
        address: { type: String, required: true }
    }]
}, { timestamps: true })

const userModel = mongoose.models.user || mongoose.model('user', userSchema)

export default userModel

File: microservices/frontend/mobile/lib/data/models/user_model.dart
/// User model — matches the exact JSON shape returned by
/// GET /api/user/data → response.userData
class UserModel {
  final String id;
  final String name;
  final String? email;
  final String? phone;
  final String role;
  final String? district;
  final List<String> crops;
  final bool isAccountVerified;
  final String? language;
  final String? preferredLanguage;
  final bool hasCompletedTour;
  final bool hasCompletedSurvey;
  final bool simpleMode;
  final List<UserAddress> addresses;

  const UserModel({
    required this.id,
    required this.name,
    this.email,
    this.phone,
    this.role = 'user',
    this.district,
    this.crops = const [],
    this.isAccountVerified = false,
    this.language,
    this.preferredLanguage,
    this.hasCompletedTour = false,
    this.hasCompletedSurvey = false,
    this.simpleMode = false,
    this.addresses = const [],
  });

  factory UserModel.fromJson(Map<String, dynamic> json) {
    return UserModel(
      id: json['id'] ?? '',
      name: json['name'] ?? '',
      email: json['email'],
      phone: json['phone'],
      role: json['role'] ?? 'user',
      district: json['district'],
      crops: json['crops'] != null
          ? List<String>.from(json['crops'])
          : [],
      isAccountVerified: json['isAccountVerified'] ?? json['is_account_verified'] ?? false,
      language: json['language'],
      preferredLanguage: json['preferredLanguage'] ?? json['preferred_language'],
      hasCompletedTour: json['hasCompletedTour'] ?? json['has_completed_tour'] ?? false,
      hasCompletedSurvey: json['hasCompletedSurvey'] ?? json['has_completed_survey'] ?? false,
      simpleMode: json['simpleMode'] ?? json['simple_mode'] ?? false,
      addresses: json['addresses'] != null
          ? (json['addresses'] as List).map((a) => UserAddress.fromJson(a)).toList()
          : [],
    );
  }

  Map<String, dynamic> toJson() => {
    'id': id,
    'name': name,
    'email': email,
    'phone': phone,
    'role': role,
    'district': district,
    'crops': crops,
    'isAccountVerified': isAccountVerified,
    'language': language,
    'preferredLanguage': preferredLanguage,
    'hasCompletedTour': hasCompletedTour,
    'hasCompletedSurvey': hasCompletedSurvey,
    'simpleMode': simpleMode,
  };

  bool get isAdmin => role == 'admin';
  bool get isFieldOfficer => role == 'field-officer';
  bool get isFarmer => role == 'user';
}

/// User address — from user_addresses table
class UserAddress {
  final String? id;
  final String fullName;
  final String phone;
  final String address;

  const UserAddress({
    this.id,
    required this.fullName,
    required this.phone,
    required this.address,
  });

  factory UserAddress.fromJson(Map<String, dynamic> json) {
    return UserAddress(
      id: json['id'],
      fullName: json['full_name'] ?? json['fullName'] ?? '',
      phone: json['phone'] ?? '',
      address: json['address'] ?? '',
    );
  }

  Map<String, dynamic> toJson() => {
    'fullName': fullName,
    'phone': phone,
    'address': address,
  };
}
